Add a discovery mechanism for nodes based on their capabilities (#59)
polkadot-fellows/RFCs: text/0059-nodes-capabilities-discovery.md

# RFC-0059: Add a discovery mechanism for nodes based on their capabilities

|                 |                                                                                             |
| --------------- | ------------------------------------------------------------------------------------------- |
| **Start Date**  | 2023-12-18                                                                                  |
| **Description** | Nodes having certain capabilities register themselves in the DHT to be discoverable         |
| **Authors**     | Pierre Krieger                                                                              |

## Summary

> This RFC proposes to make the mechanism of RFC #8 more generic by introducing the concept of "capabilities".

Reviewer: ```suggestion
This RFC proposes to make the mechanism of [RFC #8](https://github.com/polkadot-fellows/RFCs/blob/main/text/0008-parachain-bootnodes-dht.md) more generic by introducing the concept of "capabilities".
```

Implementations can implement certain "capabilities", such as serving old block headers or being a parachain bootnode.

The discovery mechanism of RFC #8 is extended to be able to discover nodes of specific capabilities.

## Motivation

> The Polkadot peer-to-peer network is made of nodes. Not all these nodes are equal. Some nodes store only the headers of recently blocks, some nodes store all the block headers and bodies since the genesis, some nodes store the storage of all blocks since the genesis, and so on.

Reviewer 2: ```suggestion
The Polkadot peer-to-peer network is made of nodes. Not all these nodes are equal. Some nodes store only the headers of recent blocks, some nodes store all the block headers and bodies since the genesis, some nodes store the storage of all blocks since the genesis, and so on.
```

It is currently not possible to know ahead of time (without connecting to it and asking) which nodes have which data available, and it is not easily possible to build a list of nodes that have a specific piece of data available.

If you want to download for example the header of block 500, you have to connect to a randomly-chosen node, ask it for block 500, and if it says that it doesn't have the block, disconnect and try another randomly-chosen node.
In certain situations such as downloading the storage of old blocks, nodes that have the information are relatively rare, and finding through trial and error a node that has the data can take a long time.

This RFC attempts to solve this problem by giving the possibility to build a list of nodes that are capable of serving specific data.

## Stakeholders

Low-level client developers.
People interested in accessing the archive of the chain.

## Explanation

*Reading RFC #8 first might help with comprehension, as this RFC is very similar.*

Please keep in mind while reading that everything below applies for both relay chains and parachains, except mentioned otherwise.

### Capabilities

> This RFC defines a list of so-called **capabilities**:

Reviewer 2: worth clarifying that this is non-exhaustive and that future capabilities may be added.

- **Archive provider**. This capability is a superset of **History provider**. In addition to the requirements of **History provider**, an implementation with this capability must be able to serve call proofs and storage proof requests of any block since the genesis up until and including their currently finalized block.
- **Parachain bootnode** (only for relay chains). An implementation with this capability must be able to serve the network request described in RFC 8.

In the context of the *head of chain provider*, the word "recent" means: any not-finalized-yet block that is equal to or an ancestor of a block that it has announced through a block announce, and any finalized block whose height is superior to its current finalized block minus **16**.
This does *not* include blocks that have been pruned because they're not a descendant of its current finalized block. In other words, blocks that aren't a descendant of the current finalized block can be thrown away.

> A gap of blocks is required due to race conditions: when a node finalizes a block, it takes some time for its peers to be made aware of this, during which they might send requests concerning older blocks. The exact gap is arbitrary.

Reviewer 2: ```suggestion
A gap of blocks is required due to race conditions: when a node finalizes a block, it takes some time for its peers to be made aware of this, during which they might send requests concerning older blocks. The exact gap is a result of network asynchrony and computational overhead.
```

Author: What I meant is that the constant "16" has been picked arbitrarily

Substrate is currently by default a **head of chain provider** provider. After it has finished warp syncing, it downloads the list of old blocks, after which it becomes a **history provider**.

> If Substrate is instead configured as an archive node, then it downloads the state of all blocks since the genesis, after which it becomes an **archive provider**, **history provider**, and **head of chain provider**.

Reviewer 2: point of information: though it doesn't matter for the purposes of this PR how the archive data is acquired, a typical node computes the incremental states rather than downloading them, no?

Author: Yes, that's maybe a poor choice of wording. I used the word "download" because it downloads the block headers and bodies.

If blocks pruning is enabled and the chain is a relay chain, then Substrate unfortunately doesn't implement any of these capabilities, not even **head of chain provider**. This is considered as a bug that should be fixed, see <https://github.com/paritytech/polkadot-sdk/issues/2733>.

### DHT provider registration

> This RFC heavily relies on the functionalities of the Kademlia DHT already in use by Polkadot. You can find a link to the specification [here](https://github.com/libp2p/specs/tree/master/kad-dht).

Reviewer 2: A permalink would be more, well, permanent

## Performance, Ergonomics, and Compatibility

### Performance

The DHT mechanism generally has a low overhead, especially given that publishing providers is done only every 24 hours.

Doing a Kademlia iterative query then sending a provider record shouldn't take more than around 50 kiB in total of bandwidth for the parachain bootnode.

Assuming 1000 nodes with a specific capability, the 20 Polkadot full nodes corresponding to that capability will each receive a sudden spike of a few megabytes of networking traffic when the `key` rotates. Again, this is relatively negligible. If this becomes a problem, one can add a random delay before a node registers itself to be the provider of the `key` corresponding to `BabeApi_next_epoch`.

Maybe the biggest uncertainty is the traffic that the 20 Polkadot full nodes will receive from light clients that desire knowing the nodes with a capability. If this every becomes a problem, this value of 20 is an arbitrary constant that can be increased for more redundancy.

### Ergonomics

Irrelevant.

### Compatibility

Irrelevant.

## Prior Art and References

Unknown.

## Unresolved Questions

> While it fundamentally doesn't change much to this RFC, using `BabeApi_currentEpoch` and `BabeApi_nextEpoch` might be inappropriate. I'm not familiar enough with good practices within the runtime to have an opinion here. Should it be an entirely new pallet?

Reviewer 2: elaborate?

Author: Maybe we could introduce two runtime functions named for example `NetworkDht_currentRandomness` and `NetworkDht_nextRandomness` that return just the randomness used in the DHT.

From a purely technical point of view, this would be exactly the same as using `BabeApi_currentEpoch` and `BabeApi_nextEpoch`, but it would be more flexible. For example, once we switch to Sassafras, the upgrade would be more seamless than if the usage of Babe is hardcoded.

As the text says, I don't have an opinion here.
